Detectors/PHOS/calib/src/PHOSBadMapCalibDevice.cxx

```diff
@@ -9,372 +9,414 @@
-      continue;
-    }
-    std::vector<uint32_t>* digits;
-    f.GetObject("Digits", digits);
-
-    std::vector<uint32_t>::const_iterator digIt = digits->cbegin();
-    std::vector<uint32_t>::const_iterator digEnd = digits->cend();
-    Cluster clu;
-    bool isNextNewEvent;
-    mBuffer->startNewEvent(); // mark stored clusters to be used for Mixing
-    while (nextCluster(digIt, digEnd, clu, isNextNewEvent)) {
-
-      fillTimeMassHisto(clu);
-      if (isNextNewEvent) {
-        mBuffer->startNewEvent(); // mark stored clusters to be used for Mixing
-      }
-    }
-    delete digits;
-    f.Close();
+  LOG(info) << "[PHOSBadMapCalibDevice - endOfStream]";
+  // calculate stuff here
+  if (calculateBadMap()) {
+    checkBadMap();
+    sendOutput(ec.outputs());
   }
 }
 
-bool PHOSCalibCollector::nextCluster(std::vector<uint32_t>::const_iterator digitIt, std::vector<uint32_t>::const_iterator digitEnd,
-                                     Cluster& clu, bool& isNextNewEvent)
+void PHOSBadMapCalibDevice::sendOutput(DataAllocator& output)
 {
-  //Scan digits belonging to cluster
-  // return true if cluster read
-  isNextNewEvent = false;
-  clu.reset();
-  int cluIndex = -1;
-  while (digitIt != digitEnd) {
-    CalibDigit d = {*digitIt};
-    if (d.mAddress == 16383) {         //impossible address, marker of new event
-      if (clu.getMultiplicity() > 0) { //already read cluster: calculate its parameters; do not go into next event
-        isNextNewEvent = true;
-        break;
-      }
-      //If just started cluster, read new event. Two first words event header
-      EventHeader h = {*digitIt};
-      mEvBC = h.mBC; //current event BC
-      digitIt++;
-      mEvOrbit = *digitIt; //current event orbit
-      digitIt++;
-      continue;
+
+  // extract CCDB infos and calibration objects, convert it to TMemFile and send them to the output
+  if (mUpdateCCDB || mForceUpdate) {
+    // prepare all info to be sent to CCDB
+    std::string flName = o2::ccdb::CcdbApi::generateFileName("BadMap");
+    std::string kind;
+    if (mMode == 0) { // Occupancy
+      kind = "PHS/Calib/BadMapOcc";
     }
-    if (cluIndex == -1) { //start new cluster
-      cluIndex = d.mCluster;
-    } else {
-      if (cluIndex != d.mCluster) { //digit belongs to text cluster: all digits from current read
-        break;
-      }
+    if (mMode == 1) { // Chi2
+      kind = "PHS/Calib/BadMapChi";
     }
-    short absId = d.mAddress;
-    float energy = d.mAdcAmp * mCalibParams->getGain(absId);
-    if (!d.mHgLg) {
-      energy *= mCalibParams->getHGLGRatio(absId);
+    if (mMode == 2) { // Pedestals
+      kind = "PHS/Calib/BadMapPed";
     }
-    clu.addDigit(absId, energy, 0., 0, 1.);
-    digitIt++;
-  }
-  if (digitIt == digitEnd && clu.getMultiplicity() == 0) {
-    return false;
+    std::map<std::string, std::string> md;
+    o2::ccdb::CcdbObjectInfo info(kind, "BadMap", flName, md, mRunStartTime, 99999999999999);
```

Comment:
Do you really need an infinite validity? Can it be more reasonable?

Reply:
Idea is to calculate bad map from current time till the next calculation. For some methods (pedestal, chi2) this can be rare: once per week or even month, for occupancy ~ once per fill. I assume, that after next iteration a more fresh object will overwrite current one, but it is hard to predict when it happen.